mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/blob/BlobStoreDeletedMessageVault.java

```diff
@@ -101,7 +102,28 @@
 
     @Override
     public Publisher<Void> delete(User user, MessageId messageId) {
-        throw new NotImplementedException("Will be implemented later");
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(messageId);
+
+        return Flux.from(messageMetadataVault.listRelatedBuckets())
+            .flatMap(bucketName -> deleteFromBucket(bucketName, user, messageId))
+            .subscribeOn(Schedulers.elastic());
+    }
+
+    private Mono<Void> deleteFromBucket(BucketName bucketName, User user, MessageId messageId) {
+        return Flux.from(messageMetadataVault.listMessages(bucketName, user))
+            .filter(deletingInformation -> messageId.equals(deletingInformation.getDeletedMessage().getMessageId()))
+            .flatMap(deletingInformation -> deleteMessageData(bucketName, deletingInformation)
+                .then(deleteMessageMetaData(bucketName, user, messageId)))
```

Comment:
I don't think that this `then` call needs to be nested?

Reply:
If the upstream is Mono, then I can do it, but here the upstream is Flux, I want to centralize the `deleting message` operation into one Mono rather spread it in Flux pipelines